### PR TITLE
Input wise masks for mask gradients

### PR DESCRIPTION
**Pull request by oliveradk**
Adds support for computing input-wise mask gradients. Useful for e.g. doing anomaly detection using edge attribution scores

### CODE REVIEW

**Comment by oliveradk** [comment]
TODO: factor out other changes

File: auto_circuit/data.py [hunk 1]
```diff
@@ -94,6 +94,7 @@ def __init__(
         kv_cache: Optional[HookedTransformerKeyValueCache] = None,
         seq_labels: Optional[List[str]] = None,
         word_idxs: Dict[str, int] = {},
+        collate_fn = collate_fn,
```
Comment (UFO-101) [inline auto_circuit/data.py]: Looks like this isn't being used?

File: auto_circuit/utils/graph_utils.py [hunk 2]
```diff
@@ -307,6 +308,10 @@ def patch_mode(
         wrapper.curr_src_outs = curr_src_outs
         if wrapper.is_dest:
             wrapper.patch_src_outs = patch_src_outs
+            if batch_size is not None:
+                wrapper.set_patch_mask_size(batch_size)
+                wrapper.patch_mask_batch.detach_().requires_grad_(True)
```
Comment (UFO-101) [inline auto_circuit/utils/graph_utils.py]: This `requires_grad` be set in `train_mask_mode`, not here.

File: auto_circuit/utils/patchable_model.py [hunk 3]
```diff
@@ -129,6 +129,16 @@ def __init__(
...
         self.wrapped_model = wrapped_model
+    
+    def set_patch_masks(self, batch_size: int):
+        self.batch_size = batch_size 
+        # update patch masks in modules
+        for dest_wrapper in self.dest_wrappers:
+            dest_wrapper.set_patch_mask_size(batch_size)
+        
+        # update patch masks in the model
+        for dest_wrapper in self.dest_wrappers:
```
Comment (UFO-101) [inline auto_circuit/utils/patchable_model.py]: These can be combined into a single for loop

**Review by UFO-101 — CHANGES_REQUESTED** [review]
Thanks for the PR! This is a nice idea, although I think executing it in a clean way is somewhat challenging.

I'm nervous about a couple of things here
- It's not ideal to create a new `patch_mask_batch` parameter. It seems this should be integrated into the existing `mask` parameter. Changing the `mask` parameter for individual edges would no longer work because the main `mask` parameter would no longer be used.
- But some existing functions will break if you just overwrite the `mask` parameter with the new batch dimension (because edge indices will no longer work.)
- The `patch_mode` context manager shouldn't permanently alter the state of the model.

I would propose making this change in two steps
1. Add a new dummy batch dimension (with size `1`) on to the `mask` parameters that is always present. Update the edge index functions to work as before with this. The core logic in PatchWrapperImpl can stay mostly the same - we will just broadcast the batch dimension by default. (and update `sample_hard_concrete` to not add a new dimension.)
2. Add a new context manager in `graph_utils`, `set_mask_batch_size` that temporarily adjusts the size of the mask batch dimension, allowing you to get the gradients for each batch element separately.

**Comment by oliveradk** [comment]
Thanks for the feedback! I mostly implemented your suggestions, but instead of adding a dimension to patch_mask by default, only add the batch dimension if `set_mask_batch_size` is called - this minimizes the chance of any downstream bugs being introduced, and I don't think any of the edge indexing functionality is critical if the main use case is to collect attribution scores over batches (please let me know if I'm missing something crucial there)

File: auto_circuit/utils/tensor_ops.py [hunk 4]
```diff
@@ -9,20 +9,26 @@
...
 
     Returns:
         A sample for each element in the mask for each batch element. The returned
         tensor has shape `(batch_size, *mask.shape)`.
     """
-    mask = mask.repeat(batch_size, *([1] * mask.ndim))
+    if not mask_expanded:
+        mask = mask.repeat(batch_size, *([1] * mask.ndim))
+    else:
+        assert mask.size(0) == batch_size
```
Comment (UFO-101) [inline auto_circuit/utils/tensor_ops.py:31]: Instead of adding a new parameter, it would be better to just check if it's already the correct shape and adjust accordingly.

File: auto_circuit/utils/patch_wrapper.py [hunk 5]
```diff
@@ -85,11 +85,29 @@ def __init__(
             self.mask_fn: MaskFn = None
             self.dropout_layer: t.nn.Module = t.nn.Dropout(p=0.0)
         self.patch_mode = False
+        self.batch_size = None
 
         assert head_dim is None or seq_dim is None or head_dim > seq_dim
         dims = range(1, max(head_dim if head_dim else 2, seq_dim if seq_dim else 2))
         self.dims = " ".join(["seq" if i == seq_dim else f"d{i}" for i in dims])
 
+    def set_mask_batch_size(self, batch_size: int | None):
```
Comment (UFO-101) [inline auto_circuit/utils/patch_wrapper.py:94]: Please add docstring. With warning that this should not be used directly, but with context manager.

You can insert a link to the context manager in the docstring using this notation:
[`set_mask_batch_size`][auto_circuit.utils.graph_utils.set_mask_batch_size]

File: auto_circuit/utils/graph_utils.py [hunk 6]
```diff
@@ -395,6 +395,28 @@ def mask_fn_mode(model: PatchableModel, mask_fn: MaskFn, dropout_p: float = 0.0)
...
+    """
+    Context manager to set the batch size of the patch masks in the model.
+
+    Args:
+        model: The patchable model to alter.
+        batch_size: The batch size to set the patch masks to. If `None`, the batch size
+            is not modified.
+
+    Warning:
+        This function modifies the state of the model! This is a likely source of bugs.
```
Comment (UFO-101) [inline auto_circuit/utils/graph_utils.py:414]: This needs another warning. That it breaks other functions of the library while the context is active and should be considered an experimental feature.

File: tests/utils/test_instance_grads.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    patch_outs = {}
+    for batch in train_loader:
+        patch_outs[batch.key] = src_ablations(
+            model, batch.clean, ablation_type=AblationType.ZERO
+        )
+
+    # compute prune scores in inputs
+    prune_scores: Dict[str, List[t.Tensor]] = defaultdict(list)
+    with set_mask_batch_size(model, batch_size):
+        with train_mask_mode(model):
```
Comment (UFO-101) [inline tests/utils/test_instance_grads.py]: Merge with statements.

File: tests/utils/test_instance_grads.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    patch_mode,
+    set_all_masks,
+    set_mask_batch_size,
+    train_mask_mode,
+)
+from auto_circuit.utils.tensor_ops import batch_avg_answer_diff
+
+
+def test_instance_grads(
+    mini_tl_transformer: HookedTransformer, dataset_name: str = "mini_prompts"
```
Comment (UFO-101) [inline tests/utils/test_instance_grads.py]: dataset_name does not need to be a parameter?

File: tests/utils/test_instance_grads.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    )
+    model = task.model
+    train_loader = task.train_loader
+
+    # compute patch outs
+    patch_outs = {}
+    for batch in train_loader:
+        patch_outs[batch.key] = src_ablations(
+            model, batch.clean, ablation_type=AblationType.ZERO
+        )
```
Comment (UFO-101) [inline tests/utils/test_instance_grads.py]: patch_outs is the same for all batches as this is zero ablation, so this loop is not required

File: tests/utils/test_instance_grads.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+    # compute prune scores in inputs
+    prune_scores: Dict[str, List[t.Tensor]] = defaultdict(list)
```
Comment (UFO-101) [inline tests/utils/test_instance_grads.py]: Should use existing PruneScores type

File: tests/utils/test_instance_grads.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+            for batch in train_loader:
+                patch_src_outs = patch_outs[batch.key].clone().detach()
+                with patch_mode(model, patch_src_outs):
+                    logits = model(batch.clean)[model.out_slice]
+                    loss = -batch_avg_answer_diff(logits, batch)
+                    loss.backward()
+                for dest_wrapper in model.dest_wrappers:
+                    prune_scores[dest_wrapper.module_name].append(
+                        dest_wrapper.patch_mask.grad.detach().clone()
```
Comment (UFO-101) [inline tests/utils/test_instance_grads.py]: append is not required. each dest_wrapper has only one patch_mask (which is why `PruneScores` type is just `Dict[str, t.Tensor]`

File: tests/utils/test_instance_grads.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+            set_all_masks(model, val=0.0)
+
+            for batch in train_loader:
+                patch_src_outs = patch_outs[batch.key].clone().detach()
+                with patch_mode(model, patch_src_outs):
+                    logits = model(batch.clean)[model.out_slice]
+                    loss = -batch_avg_answer_diff(logits, batch)
+                    loss.backward()
+                for dest_wrapper in model.dest_wrappers:
+                    prune_scores[dest_wrapper.module_name].append(
```
Comment (UFO-101) [inline tests/utils/test_instance_grads.py]: Add assertion for the mask sizes

File: tests/utils/test_instance_grads.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+                        dest_wrapper.patch_mask.grad.detach().clone()
+                    )
+                model.zero_grad()
+
+    ex_prune_score = next(iter(prune_scores.values()))[0]
```
Comment (UFO-101) [inline tests/utils/test_instance_grads.py]: Assert that each batch element mask has unique gradients

**Review by UFO-101 — CHANGES_REQUESTED** [review]
Thanks for the changes! I think this a big improvement.

I'm still not super happy with this implementation as it breaks a bunch of other features. But I recognize that a more general solution is a bigger project which is not relevant to your usecase so I don't want to block this too much longer.

Main requests are to add a warning about these problems in the docstrings. And to improve the test a little bit.

File: auto_circuit/utils/patch_wrapper.py (hunk 5 above)
Comment (UFO-101) [reply auto_circuit/utils/patch_wrapper.py:94]: And probably also include the warnings about this breaking various features of the library and to be used with caution etc.

File: auto_circuit/utils/tensor_ops.py (hunk 4 above)
Comment (oliveradk) [reply auto_circuit/utils/tensor_ops.py:31]: some of the masks are 1d though, so unclear how to distinguish an expanded 1d mask and a 2d mask where the first dimension happens to equal the batch size

File: tests/utils/test_instance_grads.py (hunk 11 above)
Comment (oliveradk) [reply tests/utils/test_instance_grads.py]: but we're accumulating prune scores across batches (so for each dest_wrapper there will be a list of batches of prune scores (this isn't strictly necessary for the test, but doubles as demo code for how to use the context handler)

File: auto_circuit/utils/tensor_ops.py (hunk 4 above)
Comment (UFO-101) [reply auto_circuit/utils/tensor_ops.py:31]: Ah yes, good point. Seems fine how it is then.

File: tests/utils/test_instance_grads.py (hunk 11 above)
Comment (UFO-101) [reply tests/utils/test_instance_grads.py]: I see, maybe add a comment with that.

File: tests/utils/test_instance_grads.py (hunk 10 above)
Comment (oliveradk) [reply tests/utils/test_instance_grads.py]: collecting a list of score batches - I guess this is not strictly necessary for this test b/c only using one batch - I changed the number of batches to 2 to better reflect the use case

File: tests/utils/test_instance_grads.py (hunk 11 above)
Comment (oliveradk) [reply tests/utils/test_instance_grads.py]: added a comment, renamed to prune_scores_batches, and concatenated afterwords into prune_scores (with proper type) to make more clear

**Review by UFO-101 — APPROVED** [review]
Looks great, thanks for this!

**Comment by UFO-101** [comment]
I will just do a couple of checks locally and then merge in next hour or so

**Comment by UFO-101** [comment]
@oliveradk Could you please fix merge conflicts and final comment? Then I will merge.

File: tests/utils/test_instance_grads.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        _dataset_name="mini_prompts"
+    )
+    model = task.model
+    train_loader = task.train_loader
+
+    # compute patch outs
+    zero_out = src_ablations(
+        model, next(iter(train_loader)).clean, ablation_type=AblationType.ZERO
+    )
+    patch_outs = {batch.key: zero_out.clone() for batch in train_loader}
```
Comment (UFO-101) [inline tests/utils/test_instance_grads.py]: This dictionary is unnecessary just pass `zero_out` directly to `patch_mode`

**Comment by oliveradk** [comment]
Fixed merged conflicts and addressed comment (I had been running the test incorrectly and had to make some more tweaks)
